### ISSUE TITLE
Sync high CPU and memory consumption

### ISSUE THREAD

**Issue by kiyanovsky** [pouchdb/pouchdb#6421]
### Issue
During initial synchronization, memory peak consumption is +500MB, CPU load - 25% during 3 mins. 
99% of downloading documents are deleted documents.
During this high load, PouchDB operations are frozen.
DB size - 209KB, number of documents: 23, number of deleted documents: 1249.

I've checked the following and it **doesn't make any difference**:
- adapters - web-sql, memory, worker (this doesn't block DOM)
- batch_size - 100/1000
- auto compaction - false/true
- poudbdb 6.0.5/6.1.2

### Info
- Environment: browser extension.
- Platform: Chrome
- Adapter: IndexedDB, WebSQL, Memory
- Server: Local CouchDB 2.0.0 on Windows 10

### Reproduce
Just sync() with live and retry = true. I can provide an export of the DB made by pouchdb.replication-stream.

**Comment by angkec** [comment]
Experiencing this issue too. Platform Safari seems to take a bigger CPU hit, one core is totally consumed and the react UI choppy.

**Comment by rvanzon** [comment]
I have the same problem. MacOS, CouchDB 2.0, Chrome 58.0.3029.110. 

Problem occurs with and without `retry` and `live` = true. There doesn't seem to be any network traffic.

**Comment by rvanzon** [comment]
It seems to happen the first time PouchDB syncs (so after deleting the IndexedDB).

I made a video of what happens. It contains Activity Monitor + Chrome Console. The webapp is reloaded and PouchDB starts to Sync. CPU goes to 120%. Around 3:07 it continues  to load some local-documents and CPU goes down.

https://www.youtube.com/edit?o=U&video_id=tP0l85LQElo

(btw, from 0:05 to 0.08 you can see that the CPU usage of beam.smp of CouchDB goes to 204%. This kinda concerns me too...)

**Comment by kiyanovsky** [comment]
I've just checked PouchDB 6.3.4 and can confirm that this issue is no longer appears. During initial sync of ~ 4000 deleted docs, memory consumption doesn't group significantly, CPU goes up to ~ 25% not for short time, db is not being locked.

**Comment by JeevanJain** [comment]
I am using updated pouchdb and pouchdb-find updated packages ... still high memory consumption..tried with one time syncing when app starting...

**Comment by julianhille** [comment]
Same here:
pouchdb 6.3.4 or 6.4.3 using leveldown in nodejs
initial sync of 4000 docs goes up to 600MB on initial sync
batch size is set to 10 batch limit is set to 5. 

As far as i got is:
- its inside the batching from a remote machine and cause by a bigger number of revs. (big is around 20 per doc)

**Comment by kiyanovsky** [comment]
For those, who are still experiencing this issue, the following can help.

The root cause of the high memory and CPU consumption - many leaf (conflict) revisions in the existing documents and those synchronized from the Cloud. When PouchDB parses JSON, a complex structure can easily take 4GB+ memory with high CPU usage. In my case, my browser extension gets killed when reaching 4GB memory limit, making synchronization impossible. You can read some additional information in [this tread](https://news.ycombinator.com/item?id=13103026).

After making some experiments, I concluded the following.

1. Process conflicts and delete non-winning revisions. It will end up with dead leaf revisions, which will be addressed below.
2. To avoid replication of dead leaf revisions, use the 'main_only' value for the options.style in replication. In this case, even if a document has many dead leaf revisions, they will not be synchronized from the Cloud to a local DB. For one of my users, this option alone resolved an issue with sync taking 4GB+ and throwing an exception, resulting in just 1GB of memory consumption and successful sync. Non-deleted conflict revisions will be replicated with this option. To be honest, I don't understand why is this not by default. Why on earth I would need to replicate to a local DB dead leaf revisions.
3. One of my tests was to read all local docs with all leaf revisions. It was failing for a single doc with many dead leaf revisions (4GB+ memory and exception). After I compacted the DB using db.compact(), the test was able to read all revisions. My experiments with auto-compaction did not go well, so I'm just running db.compact() during idle time.
4. To avoid accumulation of dead leaf revisions, you can purge them using .purge() function. It works only with IndexedDB adapter and doesn't work in a worker.

To summarize on how to resolve high memory and CPU usage:
1. Resolve and delete conflict revisions.
2. Sync from Cloud only winning revisions using {style: 'main_only'} option.
3. Compact local DB periodically using db.compact().
4. Optimize local DB by deleting dead leaf revisions periodically using db.purge().
